### ISSUE TITLE
ECH (Encrypted client hello) support

### ISSUE THREAD

**Issue by Tristan971** [haproxy/haproxy#1924]
### Your Feature Request

ECH (Encrypted client hello) is a developing specification for encrypting the original client hello in an HTTP1.1/HTTP2 context.

The main purpose is making SNI sniffing impossible by middle boxes and other such adversarial systems. The practical implementation is not dissimilar to how SSL certificates trust is established, by using certain new DNS records (+ DNS-over-HTTPS/TLS) as source of public keys involved (instead of a few CAs).

I had a short chat on the topic with @wlallemand at HAProxyConf, and he was aware of it and of the PoC referenced below. He hinted at it maybe being less relevant than before due to QUIC bringing encryption all the way through, but QUIC reaching the same level of usage as HTTP1.1/2 will take years. Especially when it still relies on Alt-Svc response headers at the moment, and while one will soon be able to advertise QUIC at the DNS level directly (see https://datatracker.ietf.org/doc/draft-ietf-dnsop-svcb-https/) this is also going to take a hot minute to be widely available, so I'm still quite interested in ECH in general (and hopefully I'm not alone in that).

Some relevant references/notes:
- IETF draft: https://datatracker.ietf.org/doc/draft-ietf-tls-esni/
- Cloudflare article on the topic: https://blog.cloudflare.com/encrypted-client-hello/ and public rollout announcement: https://blog.cloudflare.com/announcing-encrypted-client-hello/
- PoC implementation for HAProxy (and other popular webservers/load-balancer software): https://defo.ie/ (and specifically https://github.com/sftcd/openssl/blob/ECH-draft-13a/esnistuff/haproxy.md)
- Browser tracking bug for Chromium: https://bugs.chromium.org/p/chromium/issues/detail?id=1091403 and feature now shipping to stable in version 117: https://chromestatus.com/feature/6196703843581952
- Browser tracking bug for Firefox: ~https://bugzilla.mozilla.org/show_bug.cgi?id=1590863~ https://bugzilla.mozilla.org/show_bug.cgi?id=1725938
- No tracking bug for Webkit since it's more constrained in scope (as a rendering engine rather than a fully-fledged browser per se, see https://github.com/WebKit/standards-positions/issues/46)

More specifically for HAProxy, the work done by the DEfO PoC people has progressed quite a bit on the OpenSSL side:
- ECH tracking issue https://github.com/openssl/openssl/issues/7482
- Which itself is waiting on HPKE support (issue https://github.com/openssl/openssl/issues/14748) but the relevant PR for that issue is getting to its final stages so hopefully getting close 🤞 (the PR author is the same person that did the early exploration of ECH on defo.ie)

This is still somewhat early days (need HPKE merged, ECH to go from draft to RFC, and OpenSSL to adopt ECH) but I thought I'd raise this issue to have it in the tracker.

### What are you trying to do?

Use ECH with HAProxy

### Output of `haproxy -vv`

```plain
HAProxy version 2.7-dev8-7941ead+mangadex-cd2a7ce 2022-11-01T14:10+00:00 - https://haproxy.org/
Status: development branch - not safe for use in production.
Known bugs: https://github.com/haproxy/haproxy/issues?q=is:issue+is:open
Running on: Linux 5.4.143-1-pve #1 SMP PVE 5.4.143-1 (Tue, 28 Sep 2021 09:10:37 +0200) x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -ggdb3 -gdwarf-4 -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wnull-dereference -fwrapv -Wno-unknown-warning-option -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment -DMAX_SESS_STKCTR=5
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_STATIC_PCRE2=1 USE_LIBCRYPT=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_TFO=1 USE_NS=1 USE_SYSTEMD=1 USE_QUIC=1 USE_PROMEX=1
  DEBUG   = -DDEBUG_MEMORY_POOLS -DDEBUG_STRICT

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL +THREAD -PTHREAD_EMULATION +BACKTRACE -STATIC_PCRE +STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -ENGINE +GETADDRINFO +OPENSSL +LUA +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT +QUIC +PROMEX -MEMORY_PROFILING +SHM_OPEN

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=8).
Built with OpenSSL version : OpenSSL 1.1.1q+quic-mangadex-cd2a7ce  1 Nov 2022
Running on OpenSSL version : OpenSSL 1.1.1q+quic-mangadex-cd2a7ce  1 Nov 2022
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.6
Built with the Prometheus exporter as a service
Built with network namespace support.
Support for malloc_trim() is enabled.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.40 2022-04-14
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with clang compiler version 14.0.6

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
       quic : mode=HTTP  side=FE     mux=QUIC  flags=HTX|NO_UPG|FRAMED
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=

Available services : prometheus-exporter
Available filters :
        [BWLIM] bwlim-in
        [BWLIM] bwlim-out
        [CACHE] cache
        [COMP] compression
        [FCGI] fcgi-app
        [SPOE] spoe
        [TRACE] trace
```

**Comment by Tristan971** [comment]
And the HPKE prerequisite on OpenSSL side was just merged in https://github.com/openssl/openssl/commit/ad062480f7490197b174edad8625ce40d74f6e68 🙂

**Comment by sftcd** [comment]
Hiya,

I did the PoC referred to above. While it'll be a while before that'd be ready, there are two things on which some discussion could be useful in the meantime:

1) Are my proposed new OpenSSL API's for ECH ok for haproxy? My current attempt at that is [here](https://github.com/sftcd/openssl/blob/ECH-draft-13c/include/openssl/ech.h) - that differs a little from what's used in the PoC (I'll update the PoC sometime in the not-too-distant) and hasn't been reviewed by OpenSSL folk, but if some haproxy people had comments on those APIs, that'd be good input.

2) The ECH integration with haproxy is mostly easy, but there is one hard problem. When operating in ECH split-mode, if one encounters a TLS HelloRetryRequest (HRR) then the haproxy instance has to be able to do ECH processing (decryption) of  two ClientHello messages, but that just doesn't seem to match the haproxy architecture which (when in ``tcp mode`` which maps to ECH split-mode) assumes that haproxy only needs to examine the first message/ClientHello (which is entirely reasonable until one gets to ECH and HRR). My PoC didn't attempt to resolve that problem.

I'd be happy to chat about either of the above as I work on a PR for OpenSSL for ECH. (That won't be done in a few weeks, but maybe early in the new year is reasonable.)

**Comment by wlallemand** [comment]
Thank you Tristan, that's good news. We had discussions about this in the past but nothing was merged in master.

HAProxy isn't well suited to handle "split-mode", because it will only be used as TCP-mode, and this mode is basically a tunnel once everything is connected, and handling HTTP connections with the TCP mode is not really recommended. We could discuss this more later but I think we should focus on the "shared" mode for now, which will be more easily integrated with the haproxy architecture.

In my opinion the "shared" mode, aka HAProxy as a TLS endpoint, could be integrated in haproxy, I don't know what are the state of the patches at the moment, but tit looks like the fork is not up to date.

**Comment by Tristan971** [comment]
Seconding William in that if one REALLY wants TCP mode haproxy + ECH they can use HAProxy as a pure « dumb » TCP forwarder (and yes it likely can’t do TCP SNI routing reliably in that case) and have whatever handles the actual TLS layer support the ECH bits. (And at that point you likely ought to be using something like LVS instead but that’s a whole other recurring topic because we all like haproxy too much to accept using something else :-) )

At least in my case only the so-called shared mode is really of interest, so it’s good that it’s the easier one 😄

**Comment by Tristan971** [comment]
wrt to @sftcd:

> Are my proposed new OpenSSL API's for ECH ok for haproxy?

It's been a few billion years since I last wrote some C but I'll try and see if I can find time to trial your OpenSSL API proposal patches and get somewhere (in ~2-3 weeks likely)

**Comment by wenerme** [comment]
Chrome 117 support ECH now, any news on HAProxy ECH ?

**Comment by Tristan971** [comment]
Certainly didn't have time to look into it myself just yet, unfortunately :/ In other news, I updated the tracking bug for FF as they now migrated they meta-issue from ESNI to ECH here: https://bugzilla.mozilla.org/show_bug.cgi?id=1725938

**Comment by sftcd** [comment]
On 13/09/2023 07:35, 陈杨文 wrote:
> Chrome 117 support ECH now, any news on HAProxy ECH ?

No news wrt releases from here, but we've kept our
OpenSSL ECH work up to date and interoperating with
browsers, including our integration with haproxy.

Details at [1].

We're thinking of making some PPAs available in the
future, so if that's of interest feel free to get in
touch.

Cheers,
S.

[1] https://defo.ie/

**Comment by Tristan971** [comment]
Fyi some more updates and links (since the Chrome shipping was spotted before I did :).

- Chrome is shipping ECH for good in 117 as per https://chromestatus.com/feature/6196703843581952
- Cloudflare now updated their setup from ESNI to ECH https://blog.cloudflare.com/announcing-encrypted-client-hello/

Since wolfSSL [has support for it](https://github.com/wolfSSL/wolfssl/issues/6461#issuecomment-1572198792) (when built with HPKE support), and given the woes with OpenSSL already, is it perhaps something to consider without waiting for OpenSSL?

I ask because I am quite interested in it (and very willing to provide a testing ground for it, just like with QUIC before) and from the last few discussions on the ML and other issues here, there's less of a will to have HAProxy tied to OpenSSL than before and they seem intent on waiting until it is an RFC at least to consider it, which sftcd was guessing [here](https://github.com/openssl/openssl/issues/7482#issuecomment-1662161360) would be at best Q1 2024.

Also to note that when OpenSSL comes to support it, it will necessarily end up being a 3.1.x thing at best anyway... Which is going to be problematic wrt performance and distro availability (since the current distro generation is on 3.0.x mostly, and unlikely to ship 3.1.x before late-2024 at best...)

**Comment by Fireant456** [comment]
We are running into issues utilizing HAProxy and our CloudFlare DNS as it appears CF is enforcing ECH by default causing issues with updated Chrome users. Is there a solution or work around to this?

I am guessing HAProxy would need to support ECH or we would need to use something other than HAProxy till it is implemented?

**Comment by sftcd** [comment]
Not sure I understand the problem, but am interested - can you elaborate?

**Comment by Fireant456** [comment]
> Not sure I understand the problem, but am interested - can you elaborate?

We use CloudFlare for DNS and proxy on our production domains. I am unsure all the ins and outs of ECH but it seems there might be a DNS component that although our development subdomains are not proxied by Cloudflare ECH is still being enforced somehow. This is causing our older 2.1 instances of HAProxy to throw the error "ERR_ECH_FALLBACK_CERTIFICATE_INVALID" when a Chrome 217 client is trying to access one of those domains. I am guessing this is due to lack of ECH support in HAProxy?

**Comment by Tristan971** [comment]
If Cloudflare forcefully publishes eager ECH records (HTTPS-type)  on the zone forcefully, that sounds like it might cause issues indeed when clients trying ECH directly, even when they reach HAProxy without passing via CF.

Easiest would be to have CF proxying for your environment too, or to see if overriding HTTPS record on your development domain’s zone fixes it. But since CF forcefully appends their CAA records and similar, I have a feeling this wouldn’t quite work…

Maybe contact CF support asking them to opt-out your development domain zone from ECH. It’s likely they can do that.

**Comment by Fireant456** [comment]
I think that we are having luck with the proxy, but not sure if that will change or not. We are forced to not proxy WSS connections as CF tends to break those from time to time causing connection issues for our users. So we are seeing those subdomains affected in our production environment...

**Comment by Tristan971** [comment]
Thing is, even if HAProxy did support ECH, you would have the same issue.

ECH essentially uses 2 different certs:
- 1 outer cert, visible to network operators
- 1 inner cert, hidden from network operators

And the outer cert domain is defined by the DNS zone, and must be a valid cert (cloudflare-ech.com for Cloudflare iirc).
ECH keeps privacy by assuming that the outer cert is either non-specific (generic to all CF sites in this case) or at least somewhat innocuous.
It’d have been lovely to have a generic one anyone could use (imagine everyone using example.com as outer name/cert for example), but the cert authorities consider that evil so it won’t happen ever.

The inner cert is the site-specific one you currently have. So that one is no problem.

But naturally you wouldn’t be able to get such an outer cert yourself since you don’t control cloudflare-ech.com.

I’m not familiar enough with the spec to know how HTTPS records deal with multiple possible outer certs and priorities though, but my guess is you’d need to set it up with a higher-priority record for a domain you control and generate an outer cert for.

---

Edit: as clarified below by @sftcd, calling the identification material of an ECH exchange a "certificate" is incorrect

**Comment by Fireant456** [comment]
We can see cloudflare-ech.com being queried in the failed requests so maybe there is some mechanism there to allow our downstream certs as a fallback?

**Comment by Tristan971** [comment]
I wonder. You’ll have to bring it up with Cloudflare at this point. It’s not something HAProxy can help with directly (whether it supports ECH or not), as it’s a side effect of CF’s implementation of common zone records for their customers really.

**Comment by sftcd** [comment]
Don't have much time to comment now, but just to note that ECH does not involve use of new "certs" so the descriptions above are off-base a bit.

**Comment by Tristan971** [comment]
My bad, let's say "key-pairs-of-some-sort" then.

But unless I totally misunderstood, a relevant point is that you can't respond to an outer ClientHello requesting cloudflare-ech.com unless you have some private key material specific to that domain name, which only Cloudflare has as they publish the public portion of it on that zone, as far as I understand it.

So the effect is analoguous unless CF were to make that material public, which I'm not sure they're even allowed to (just like publishing a cert private key voluntarily is forbidden by the CAB and will result in any such cert being revoked).

**Comment by sftcd** [comment]
Sorta. With ECH a browser will emit an initial TLS message (the ClientHello) that can only be fully processed by the entity with the relevant ECH private key. As I imagine CF deploy stuff, they assume that message will be sent to them. If somehow that message is sent somewhere else, then yes, ECH decryption won't happen, and the TLS handshake will fail with a (non-informative;-) message like the one quoted above.

**Comment by Tristan971** [comment]
hmmm that will be no fun for anyone using CF for DNS only (like I do too...).

![image](https://github.com/haproxy/haproxy/assets/875533/c6bf4a2a-63ed-4a46-93fc-1cda01b3873e)

Well, it is for the greater good... time to figure something out.

**Comment by Fireant456** [comment]
This seems like a large oversight and not much education was put out to compensate for this massive change. There are TONs of people that just use CloudFlare for DNS.... I have opened a ticket and replied on 1 of 2 semi-related ECH forum posts in the CF community forums so here's hoping they have some "good" solution to give us.

**Comment by sftcd** [comment]
If you only use CF for DNS, then ECH shouldn't be in the picture. CF would have to be adding "ech=" values into your HTTPS RRs for that to happen. Are they doing that? If so, that'd be visble via ``dig`` so have you an example?

**Comment by sftcd** [comment]
For example:
```
$ dig +short https rte.ie
1 . alpn="h3,h2" ipv4hint=104.18.142.17,104.18.143.17 ipv6hint=2606:4700::6812:8e11,2606:4700::6812:8f11
```
Is for a CF customer (I think) but has no ech= and hence ECH will not be attempted.

**Comment by Tristan971** [comment]
> If you only use CF for DNS, then ECH shouldn't be in the picture. CF would have to be adding "ech=" values into your HTTPS RRs for that to happen. Are they doing that? If so, that'd be visble via `dig` so have you an example?

Most likely they do it on @ if it is set to proxy mode, and then it applies to subdomains? Because on a zone of mine I do indeed not see it yet. Tho rte.ie might be on a non-free CF plan and they're delaying the rollout for those. idk.

But if for free CF accounts people are in a situation of:
- domain.com (@) set to proxied
- dev.domain.com set to dns-only

have the second one falling back to using the HTTPS record on @?

Then again, that sounds odd since HTTPS is more akin to an A/AAAA record than anything else, but idk... maybe just browser bug atm

**Comment by sftcd** [comment]
Not sure - I'm not familiar with CF's biz model(s), sorry. Doesn't (so far) sound like a browser bug though, but one could tell for a given DNS name using dig and wireshark pretty quickly, so shouldn't be hard to figure what behaviour is happening vs. expected.

**Comment by Fireant456** [comment]
> For example:
> 
> ```
> $ dig +short https rte.ie
> 1 . alpn="h3,h2" ipv4hint=104.18.142.17,104.18.143.17 ipv6hint=2606:4700::6812:8e11,2606:4700::6812:8f11
> ```
> 
> Is for a CF customer (I think) but has no ech= and hence ECH will not be attempted.

![image](https://github.com/haproxy/haproxy/assets/1479965/450e79ef-d8b0-43ba-b333-1bc2ac03fd23)

I don't seem to be getting the same output as you. If you would like to check on your end I don't mind sharing some domains that are affected. My other developer that has been looking into this issue with me noted that the ECH `SSL_ECH_CONFI_LIST` they are seeing might be from cache. So maybe caching is our issue here since some sub domains are proxied and have ECH properly working and when the user tries to access a DNS-only domain the ECH is still trying to be queried.
![image](https://github.com/haproxy/haproxy/assets/1479965/0df81007-1502-4131-8f8e-8b5d9181cb36)

Working Proxied:
sonoransoftware.com
sonorancms.com

Not Working DNS-only:
ws.sonorancms.com (WSS only, but putting into chrome should still show the error mentioned)
cms.dev.sonoransoftware.com
git.sonoran.software

NOTE: the ws.sonorancad.com request shown in the chrome logs is now working as we have moved to a paid CF plan to disable ECH on that domain entirely. We currently are using that to have an active ticket with CF on this issue.

**Comment by Tristan971** [comment]
```
$ dig +short sonoransoftware.com https
1 . alpn="h3,h2" ipv4hint=104.21.38.120,172.67.222.151 ech=AEX+DQBB+QAgACDX9TC2tUA8p/hiOiEab+YAtdI17bmXO1p6YjCnlkiXcwAEAAEAAQASY2xvdWRmbGFyZS1lY2guY29tAAA= ipv6hint=2606:4700:3031::6815:2678,2606:4700:3037::ac43:de97
$ dig +short git.sonoransoftware.com https
1 . alpn="h3,h2" ipv4hint=104.21.38.120,172.67.222.151 ech=AEX+DQBB+QAgACDX9TC2tUA8p/hiOiEab+YAtdI17bmXO1p6YjCnlkiXcwAEAAEAAQASY2xvdWRmbGFyZS1lY2guY29tAAA= ipv6hint=2606:4700:3031::6815:2678,2606:4700:3037::ac43:de97
$ dig +short git.sonoran.software https
git.dev.sonoransoftware.com.
1 . alpn="h3,h2" ipv4hint=188.114.96.0,188.114.97.0 ech=AEX+DQBB+QAgACDX9TC2tUA8p/hiOiEab+YAtdI17bmXO1p6YjCnlkiXcwAEAAEAAQASY2xvdWRmbGFyZS1lY2guY29tAAA= ipv6hint=2a06:98c1:3120::,2a06:98c1:3121::
$ dig +short sonoran.software https
1 . alpn="h3,h2" ipv4hint=104.21.54.71,172.67.136.96 ech=AEX+DQBB+QAgACDX9TC2tUA8p/hiOiEab+YAtdI17bmXO1p6YjCnlkiXcwAEAAEAAQASY2xvdWRmbGFyZS1lY2guY29tAAA= ipv6hint=2606:4700:3035::6815:3647,2606:4700:3037::ac43:8860
```
I do see the records everywhere here, might be your dig is outdated or something

**Comment by Fireant456** [comment]
> ```
> $ dig +short sonoransoftware.com https
> 1 . alpn="h3,h2" ipv4hint=104.21.38.120,172.67.222.151 ech=AEX+DQBB+QAgACDX9TC2tUA8p/hiOiEab+YAtdI17bmXO1p6YjCnlkiXcwAEAAEAAQASY2xvdWRmbGFyZS1lY2guY29tAAA= ipv6hint=2606:4700:3031::6815:2678,2606:4700:3037::ac43:de97
> $ dig +short git.sonoransoftware.com https
> 1 . alpn="h3,h2" ipv4hint=104.21.38.120,172.67.222.151 ech=AEX+DQBB+QAgACDX9TC2tUA8p/hiOiEab+YAtdI17bmXO1p6YjCnlkiXcwAEAAEAAQASY2xvdWRmbGFyZS1lY2guY29tAAA= ipv6hint=2606:4700:3031::6815:2678,2606:4700:3037::ac43:de97
> $ dig +short git.sonoran.software https
> git.dev.sonoransoftware.com.
> 1 . alpn="h3,h2" ipv4hint=188.114.96.0,188.114.97.0 ech=AEX+DQBB+QAgACDX9TC2tUA8p/hiOiEab+YAtdI17bmXO1p6YjCnlkiXcwAEAAEAAQASY2xvdWRmbGFyZS1lY2guY29tAAA= ipv6hint=2a06:98c1:3120::,2a06:98c1:3121::
> $ dig +short sonoran.software https
> 1 . alpn="h3,h2" ipv4hint=104.21.54.71,172.67.136.96 ech=AEX+DQBB+QAgACDX9TC2tUA8p/hiOiEab+YAtdI17bmXO1p6YjCnlkiXcwAEAAEAAQASY2xvdWRmbGFyZS1lY2guY29tAAA= ipv6hint=2606:4700:3035::6815:3647,2606:4700:3037::ac43:8860
> ```
> 
> I do see the records everywhere here, might be your dig is outdated or something

Not sure, just installed it via my WSL AlmaLinux 8 environment 🤷

**Comment by sftcd** [comment]
I see the same ``dig`` outputs as the post just above, e.g.:
```
$ dig +short sonoransoftware.com https
1 . alpn="h3,h2" ipv4hint=104.21.38.120,172.67.222.151 ech=AEX+DQBB+QAgACDX9TC2tUA8p/hiOiEab+YAtdI17bmXO1p6YjCnlkiXcwAEAAEAAQASY2xvdWRmbGFyZS1lY2guY29tAAA= ipv6hint=2606:4700:3031::6815:2678,2606:4700:3037::ac43:de97
```

**Comment by Tristan971** [comment]
That said we strictly stick to DNS-only usage of CF and indeed these records don't show up yet on our zone so idk.

```
$ dig +short mangadex.org https
$
```

But since they do still generate TLS certs for our domain, and inject their own CAA records as well, I suspect that it is an oversight of theirs and we'll soon be in the same situation as you. So at this point we're just going to move out of CF for auth NS, just in case.

Nonetheless, all of this should very much be a discussion with CF.

**Comment by Fireant456** [comment]
Yep ageed, I appreciate you assisting me a bit in diagnosing this. Hopefully not downplaying this issue :)

**Comment by lukastribus** [comment]
The Cloudflare issue is indeed off topic. I guess people will be on the lookout for a workaround and then may come back here posting:

It looks like disabling TLSv1.3 in your Cloudflare settings works around this problem even on the free plan, so you avoid the need to upgrade to a paid plan just to restore DNS-only services.

Source: https://community.cloudflare.com/t/getting-ech-fallback-certificate-error/565167/16

All further discussions regarding Cloudflare belong into their community forums (or support) of course.

Let's keep this feature request on topic.

**Comment by Ralle** [comment]
Hello guys,

I used this ticket as a reference for a discussion on Cloudflare I made today. The team responded and are rolling back ECH. Thank you for helping me helping us :)

https://community.cloudflare.com/t/err-ech-fallback-certificate-invalid-with-subdomain/704703/8

**Comment by crabique** [comment]
It appears ECH is now supported by all major browsers and seems to be the final form of that solution, would be very cool to have it available in HAProxy. TLS temination mode on high-density things like CDNs and app hosting is where it would truly shine, due to the frontend to backend numbers ratio.

**Comment by awptechnologies** [comment]
I use OPNsense with haproxy. This is backed by acme certs. Since ECH has come out i get errors with certs. I use a single cert with multiple subject alt names and it seems the browsers don't understand this. After disabling ech in firefox with a custom json file it works fine. My issue is mobile firefox refuses to work. Does anyone have any idea if this is happening because of ech or the lack of ech support in haproxy? If i seperate all of the subject alt names into individual certs would that put a band-aid on the issue?

**Comment by sftcd** [comment]
FWIW, I'd be surprised were that due to ECH. haproxy doesn't currently include any ECH code in releeases AFAIK. Perhaps your 'custom json file' is doing more than disabling ECH?

**Comment by awptechnologies** [comment]
It doesn't.

Also it seams anything other then firefox doesn't have issues so idk what they have going on.

Here is contents of my json file.
`{
  "policies": {
    "DisableEncryptedClientHello": true
  }
}
`

**Comment by lukastribus** [comment]
This has nothing to do with haproxy, most of the web does not support ECH.

Instead you need to troubleshoot this between your nameserver configuration and firefox. Do your nameserver advertise HTTPS records? Do those HTTPS records show a ECH configuration?

Are you on a free plan of Cloudflare? Then you *are* subject to their beta testing.

None of this is in any way related to haproxy.

**Comment by DemiMarie** [comment]
ECH would be useful in layer 4 proxying as well as for layer 7.

**Comment by HiHat** [comment]
"To accelerate adoption, we added ECH support to [Chrome](https://chromestatus.com/feature/6196703843581952) and our [BoringSSL](https://issues.chromium.org/issues/40644929) cryptographic library, which enables apps. Chrome also collaborated with partners like Cloudflare to validate interoperability.

Now that the ECH standard is finalized and approved to be published by the IETF in the coming weeks, we expect development and deployment to accelerate towards closing this long-standing privacy gap."

https://medium.com/jigsaw/a-more-private-internet-encryption-standards-hit-new-milestones-c239ede23eaf

**Comment by sftcd** [comment]
There is ECH news to report! :-)

- ECH is not yet part of an OpenSSL release. We'd hope ECH will be part of OpenSSL 4.0 in April 2026. However, we have been working with OpenSSL maintainers on the so-called ["ECH feature branch"](https://github.com/openssl/openssl/tree/feature/ech) and that branch (subject to the same OpenSSL maintainer approval process as the OpenSSL master branch) now includes sufficient ECH code for servers like haproxy. 
- I've prepared an [ECH-shared](https://github.com/sftcd/haproxy/tree/ECH-shared) branch of haproxy that supports ECH shared-mode and that uses the OpenSSL project's ECH feature branch. That code could be discussed, improved and integrated whenever the haproxy maintainers are happy enough with it. (I need to do a bit more testing of that in the next few days.)
- The above branch supports the case where haproxy is both the ECH and the TLS endpoint.
    - We do have a branch [here](https://github.com/defo-project/haproxy) that additionally includes ECH split-mode where haproxy is the ECH endpoint but not the TLS endpoint. The APIs for that have yet to be discussed with the OpenSSL maintainers, so split-mode will be further down the road.
- There's some [howto text](https://github.com/defo-project/ech-dev-utils/blob/main/howtos/haproxy.md) that describes building, the code changes etc.
- In other related news:
     - ECH support using these OpenSSL ECH APIs was [included in the ligthttpd](https://redmine.lighttpd.net/projects/lighttpd/wiki/TLS_ECH) web server (in January 2025) so some code and patterns are common with that. 
    - We have proposed a very similar PR for the NGINX project (https://github.com/nginx/nginx/pull/840).
    - We have proposed a very similar PR for the Apache2 project (https://github.com/apache/httpd/pull/551) that's seen some good review/discussion

Lastly, for open-ness, our work on this has been funded by the Open Technology Fund (OTF) in the [DEfO project](https://defo.ie/).

**Comment by wlallemand** [comment]
Hello @sftcd, I took a look at the ECH-shared branch and made my first comments directly in the github interface.

I think we should separate things in multiple patches:

- certificate loading + ECH feature
- sample fetch / log
- commands on the stats socket.

We'll try to have this merged for the 3.3 release which is supposed to be released in November.

There are things that are uncertain for me on the way of loading this, I need to make more tests about it.

**Comment by sftcd** [comment]
Thanks for the comments! Entirely happy to split things up to suit whatever you'd like. I'll go over those comments this evening (my time) as I've a bit of $dayjob stuff to do now. We also have a bunch of test scripts that might also be useful for you in playing with this - I'll send pointers to those later too, but for sure, adding tests will be needed and isn't there yet.

**Comment by wlallemand** [comment]
@sftcd do you know the state of ECH in AWS-LC? It looks like they have an API which is not the same at all, it's probably inherited from boringSSL. I don't know if they followed the evolution of the draft or not.

**Comment by sftcd** [comment]
I think AWS-LC should inherit ECH from boring, and IIUC lighttpd supports those or OpenSSL, but I've not tested servers with either boring or AWS-LC myself. I'd have no objection to trying to make haproxy work similarly - while the OpenSSL maintainer's preferred APIs do differ a lot from boring's on the server side, we could easily copy the pattern from lighttpd and aim for both. Or, if you were ok with just doing OpenSSL first, that'd be easier for me:-)

The place to start looking at lighttpd code is [here](https://github.com/lighttpd/lighttpd1.4/blob/master/src/mod_openssl.c#L97)

**Comment by sftcd** [comment]
@wlallemand I just pushed a fixup commit that addresses the easy changes based on your comments. That include 3 `TODO:` things with bigger changes that I'll look at separately this evening/tomorrow. (Is using the `git commit --fixup` thing what you prefer btw?)

**Comment by sftcd** [comment]
Just pushed a version that moves the ECH key management code to `ssl/ech.c` as requested, so 2 `TODO:` left.

**Comment by sftcd** [comment]
next one fixes the `maxkeyfiles` issue, very simply, by just removing the limit - if more's needed happy to do more

**Comment by sftcd** [comment]
And the push just now ticks off the last `TODO:` (logging as a sample). I wouldn't say I'm confident that's done correctly though, but I'm sure @wlallemand you'll see what needs fixing:-) I think that means I've addressed all your comments from the other day (though perhaps imperfectly), so whenever you have a chance to review again that'd be great. I'll do some more testing tomorrow (on the logging and key mgmt stuff) - the basic ECH operation seems to work ok.

I think I also promised some hints about running/testing - I'm using a version of [this script](https://github.com/defo-project/ech-dev-utils/blob/main/scripts/testhaproxy.sh) - let me know if that looks helpful.

**Comment by wlallemand** [comment]
@git001 not sure what this feature request is about, we already know that ECH is supported by AWS-LC, my question was more about the state of the API regarding the current ECH draft.

@sftcd okay thank you, I'll take a  look.

**Comment by git001** [comment]
@wlallemand ah okay. Will then close the FR.

**Comment by sftcd** [comment]
@wlallemand I just pushed a better (and working:-) version of the logging thing, should be easier to say what you'd like to see in the logged string now. Next for me is to test the key mgmt stuff once that the logging seems ok.

**Comment by wlallemand** [comment]
@sftcd okay, thank you, I commented the last one.

Also I pushed a branch https://github.com/haproxy/haproxy/tree/20250916-openssl-ech with a CI job which builds OpenSSL+ECH, curl+ECH and HAProxy+ECH. I'll make a vtc file which test the feature.

**Comment by sftcd** [comment]
Would you prefer me to continue to modify my `ECH-shared` branch as we go, or that I make PRs against the one you've created? (Either is fine by me.)

**Comment by wlallemand** [comment]
Well, we don't do pull requests since the github repository is only a mirror of our official one, PR are closed by a bot and redirected to the mailing-list. So I'm fine with reviewing your branch for now.

**Comment by sftcd** [comment]
okey-dokey. will make that logging change as per your last comment and push that <strike>in a wee while</strike> after dinner

**Comment by sftcd** [comment]
pushed that logging change, now to test the key mgmt stuff...

**Comment by sftcd** [comment]
I did some testing of the ECH key mgmt stuff, and somewhat surprisingly, (haven't played with it in a good while), it seems to be working pretty fine.

The test script I made for that is [here](https://github.com/defo-project/ech-dev-utils/blob/main/scripts/socat-haproxy-test.sh).

**Comment by wlallemand** [comment]
@sftcd I saw that you changed some things in your branch, do you have more changes to push? It would be better to squash your commits together and split them in 3 commits (configuration, logs and stats sockets command).

**Comment by sftcd** [comment]
I don't think I've any outstanding changes to push, other than to do the squash then split into 3 commits as you suggest. Will try get you that tomorrow.

**Comment by sftcd** [comment]
I did a rebase with upstream, split the changes into 3 commits as requested and then force pushed. Hope that's what's wanted:-)

**Comment by wlallemand** [comment]
@sftcd I pushed a series in https://github.com/haproxy/haproxy/tree/refs/heads/20251013-wla-ech-shared  this morning. These are mostly fixes around spaces and tabs. (We use tabs to indent and spaces to align). 

I'm a bit bothered about the initialization and the way things are updated over the CLI. In HAProxy we don't update SSL_CTX directly most of the time, but we recreate them over a system of transactions. I'm experimenting with another way of initializing that, I'll update you about it.

**Comment by sftcd** [comment]
It's entirely possible the CLI stuff I did took a wrong direction. Let me know if/how I can help as you experiment.

**Comment by wlallemand** [comment]
New branch there: https://github.com/haproxy/haproxy/tree/20251023-wla-ech-shared-1 I'll squash the patches in the next branch.

I made some cleanup, moved everything in ech.c so this is less intrusive, every CLI and parser can be put there. The only thing we need outside is the loading in ssl_sock_initial_ctx()

Regarding the parsing, I removed the "ech" keyword from the crt-list, it was a bit of a misunderstanding from my side, you can't have this in a crt-list since you don't have the SNI yet at this step. So it can only be applied on the initial_ctx.

On the CLI, I started to change a few things: 
https://github.com/haproxy/haproxy/commit/116eb0a4c999dd13fc9fdef11c6eb1df52859f0b

Basically it is not a good idea to pass through every FDs in fdtab because you can have millions of them, and if you don't yield and take too long that would trigger the watchdog and kill the process. The fdtab is used for every FDs, so every connections are also there. 

What I've done is looping through the proxies_list and bind_conf of each frontend, this list don't change during runtime so there's no locking or any specific things to handle for multi-threading.

The initial_ctx is related to a bind line (bind_conf), not a frontend. You can have multiple bind in a frontend, so when updating the ECH, you probably want to specify the bind, not the frontend. I didn't address that yet but that would make more sense.

Regarding the update of an SSL_CTX directly, we could probably keep it that way, since it applies only on the initial_ctx but not on SSL_CTX associated with a SNI. Meaning there's no impact when trying to update the certificate for a SNI because the ECH is not enabled there. So we're good there.

Also I found out that the `del ssl ech` take a `[<age-in-secs>]` parameter, this is a really good idea, however this is a bit frustrating because the API does not use the age of the ECH certificate, but is related to the time we loaded the ECH in an SSL_CTX, so when reloading HAProxy we would lost the age of the certificate. I don't know if it's possible to have a generation date in the ECHConfigList from the PEM but it would make more sense to me. I might be misunderstanding the way it should work, though.

I have another branch with a few patches that integrate a tree with a cache of files for the ECH files, in order to speedup the loading when the same ECH path is used multiple time in the configuration, so it's not loaded multiple time on the disk. That's what we're doing for the other certificates everywhere in haproxy.

The feature freeze of HAProxy 3.3 is at the end of the month, I'll do what I can to have these patches merged, but I'll probably set this feature in experimental mode for this version.

**Comment by sftcd** [comment]
Great stuff. I'll take a detailed look tomorrow. Meantime:
- yes, neither the ECHConfigList nor the PEM file have date/time fields. The `age-in-secs` though isn't too bad as the expected model is to generate new keys every e.g. hour (what Cloudflare do); publish the new ones in the DNS e.g. with TTL=1800; add that to the running proxy and then delete ECH keys older than 10880 (3 hours) and `age-in-secs` matches that ok
- ECH being an experimental feature for now makes total sense, until e.g. an OpenSSL release includes it (hopefully April)

**Comment by wlallemand** [comment]
> * yes, neither the ECHConfigList nor the PEM file have date/time fields. The `age-in-secs` though isn't too bad as the expected model is to generate new keys every e.g. hour (what Cloudflare do); publish the new ones in the DNS e.g. with TTL=1800; add that to the running proxy and then delete ECH keys older than 10880 (3 hours) and `age-in-secs` matches that ok

Some people reload HAProxy a lot in order to change things in the configuration, this would reset the timer for every ECH certificates unfortunately, and won't give you any clue on which one you are supposed to remove once you reloaded.

I feel like this is the same problem as the tls-ticket-keys in haproxy. We have a static number of TLS ticket keys, the first one is used for new connections, the previous ones are used to resume connections that used the old tickets. Each time we insert a new ticket the oldest one is deleted. It's basically like a ring-buffer.

If we could set a maximum number of ECH certificates, and discard the first which was inserted once the list is full, the user just have to count the number of ECH entries it needs depending on the TTL of the DNS entry and how often he wants to add a new certificate. What do you think?

**Comment by sftcd** [comment]
The OpenSSL APIs we agreed with the maintainers for ECH don't currently support a ring buffer, sorry. That said, I suspect things as they are may be fine - so long as there aren't bazillons of ECH key pairs and there's some e.g. cronjob that deletes keys after they've been loaded for e.g .3 hours, then occasionally storing more keys than usual won't be a problem and the old ones will disappear a few hours later anyway. Thus far, nobody I've spoken to has had a good reason to want bazillions of ECH key pairs, e.g. it looks like 3 of them at a time is enough for all of cloudflare. If it's useful I can create an OpenSSL issue for the `OSSL_ECHSTORE` API to enable a ring buffer, and if that turns into merged code, you'd have that option too in future. Sound like a plan?

PS: ECH things aren't really that like X.509 certificates - they're more or less just the `public_name`, alg info and the key value, with nothing like an issuer/subject/notBefore/notAfter there at all. (Which is nice in a way - we don't have to deal with all the X.509 nonsense:-)

**Comment by sftcd** [comment]
FYI: I just built the `20251023-wla-ech-shared-1` branch and it passed my localhost tests fine.

**Comment by wlallemand** [comment]
> If it's useful I can create an OpenSSL issue for the OSSL_ECHSTORE API to enable a ring buffer, and if that turns into merged code, you'd have that option too in future. 

We can wait a bit to see what people expect, it could be enough without any limit for now!

> FYI: I just built the 20251023-wla-ech-shared-1 branch and it passed my localhost tests fine.

Okay great, I squashed the commits in this branch https://github.com/haproxy/haproxy/commits/20251029-wla-ech-shared-3/

The biggest change is the <name> format of the CLI command, which now requires the name of a bind line, instead of a frontend. We have multiple listeners by bind line and multiple bind lines per frontend.

Everything else is only some cleanup, indentation, alignment etc. I also put every keywords under the experimental-directives keyword.  I added a description on the commits as well. And started to write a documentation which I need to finish.

We're in feature freeze at the end of the week, so I plan to merge this tomorrow in master if you are fine with the changes. We can still fix bugs and add more documentation after that.
People would be able to start using it in the next dev version which is this week and report bugs about it. What do you think?

**Comment by sftcd** [comment]
> We're in feature freeze at the end of the week, so I plan to merge this tomorrow in master if you are fine with the changes. We can still fix bugs and add more documentation after that.
> People would be able to start using it in the next dev version which is this week and report bugs about it. What do you think?

Sounds perfect! Happy to help out as people play with it and improve things as we go.

**Comment by wlallemand** [comment]
@sftcd Thank you, that's merged !

**Comment by sftcd** [comment]
On 30/10/2025 09:40, William Lallemand wrote:
> wlallemand left a comment (haproxy/haproxy#1924)
> 
> @sftcd Thank you, that's merged !

Excellent! Looking forward to helping out as people use that.

**Comment by wtarreau** [comment]
and 3.3-dev11 issued with it now!

**Comment by wtarreau** [comment]
Do we need to keep this open now that it's merged ?

**Comment by sftcd** [comment]
IMO, no and thanks for processing the submission so well. If/when there're issues with using it, or a desire to extend it (e.g. for ECH split-mode), we can open a new issue. (I hope to come back with the latter in the new year-ish.)

**Comment by wtarreau** [comment]
OK fine! And thanks a lot for your long work on his point, Stephen, that's much appreciated!
